activerecord/lib/active_record/relation/batches.rb

```diff
@@ -345,7 +346,15 @@ def batch_on_loaded_relation(relation:, start:, finish:, order:, batch_limit:)
           end
         end
 
-        records.sort_by!(&:id)
+        records.sort_by! do |record|
+          unless record.id
```

Comment:
having an `id` doesn't necessarily mean that the record is persisted. let's defer to the `#persisted?` method instead which is [defined to depend on an internal attribute that tracks persistence](https://github.com/rails/rails/blob/917b6b2908510fbbc8d55db6360220485324e63c/activerecord/lib/active_record/persistence.rb#L696).

```suggestion
          unless record.persisted?
```

Reply:
in the filter block (L344-346), i see that we compare record's `id` to `start`. looks like if the id is nil, that check fails and returns a `NoMethodError`? because the `:>=` method isn't defined for nil? i believe it is another source of the comparison error but you probably didn't catch it in your test because `start` is `nil`.

Reply:
>in the filter block (L344-346), i see that we compare record's id to start.

Right, good catch. We wanted to avoid iterating over `records` just to check the presence of `id` so we decided to put the check into an existing iteration. But you are right, this still leaves a possibility of the code failing if `start` or `finish` is present. I'll make a change and add a few more test cases.

> having an id doesn't necessarily mean that the record is persisted. 

True, but it's actually intentional to check for `id` presence, or, in a bit more abstracted term - "cursor" presence. Technically all `in_batches` need to perform iterations is a cursor to be present, which in the current implementation hardcoded as `id` (primary key value) but if `in_batches` supported custom cursors it would be `record.public_send(cursor_column_name)` because technically it should be allowed to do something like
```ruby
loaded_relation.build(id: 123, text: "new record")
```
and `in_batches` will iterate over the newly built record just fine. 
But since this use-case feels unrealistic we decided to focus on a more user-friendly messaging as we expect most of the cases people will see this error when `id` is not present due to relation having a new record.

But this decision opens another corner-case, what if we use `in_batches` on a relation with no primary key and thus none of the records will have `id`, will it misleadingly raise `UnpersistedRecordError`?

So I think I'm leaning towards changing the error name and the error message towards saying "record doesn't have a cursor value" because this is what we truly care about and what prevents `in_batches` from working. Unless I'm missing something this should remove all the ambiguity around all the corner-cases

Reply:
thanks for the detailed explanation. i think renaming the error clears things up. perhaps we could make the message a little more descriptive to help with debugging. unless it’s safe to assume that anyone using `in_batches` knows and expects it to iterate over cursor values.